### PR TITLE
AGENT-265: Agent asset interface with traits

### PR DESCRIPTION
**Pull request by andfasano**
Inspired by the initial work developed by @patrickdillon in PR #5990, also this patch offers the possibility for a writable asset to customize the way its content is persisted into the currently configured assets folder.
This approach it's a tentative to limit the refactoring impact required on the existing codebase by adopting a traits-like pattern, in conjuction with the an adapter to wrap the current default behavior.
In addition, an example of how it could be used to persist the current agent ISO image is provided.

### CODE REVIEW

File: cmd/openshift-install/create.go [hunk 1]
```diff
@@ -239,6 +239,15 @@ func newCreateCmd() *cobra.Command {
 	return cmd
 }
 
+func asFileWriter(a asset.WritableAsset) asset.FileWriter {
```
[the remark this answers is not in the document]
Comment (zaneb) [reply cmd/openshift-install/create.go:242]: It seemed like putting it inside PersistToFile would make it more a part of the Asset API - you can either return something from `Files()` or implement the `FileWriter` interface. I'm relaxed though.

[the remark this answers is not in the document]
Comment (patrickdillon) [reply cmd/openshift-install/create.go:242]: > I'm relaxed though.

Me too. I think either one of these solutions is good. I do think exceptions will be rare, and we should opt for minimal impact with the option of expanding later if we need.

> you can either return something from Files() or implement the FileWriter interface.

I don't understand your point about `Files()`. Are you saying we could remove the need for `Files()` in the interface? I think that would be desirable as all assets need to implement Files(), as shown here.

**Review by patrickdillon — COMMENTED** [review]
I think this is a great solution and I have closed my earlier PR in favor of this.

**Comment by andfasano** [comment]
Thanks @patrickdillon. I've fixed all the lint errors

**Comment by patrickdillon** [comment]
This is merging into the agent branch so I'm ok with merging as is. As this is a fundamental change I would like to run more tests to exercise this, but they are not setup with this branch. 

I suppose we can do that testing when we rebase onto master.

/approve

**Pull request merged**